test/matrix/csr_kernels.cpp

```diff
@@ -0,0 +1,149 @@
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include <ginkgo/core/matrix/csr.hpp>
+
+
+#include <algorithm>
+#include <numeric>
+#include <random>
+#include <vector>
+
+
+#include <gtest/gtest.h>
+
+
+#include <ginkgo/core/matrix/dense.hpp>
+/*#include <ginkgo/core/base/array.hpp>
+#include <ginkgo/core/base/math.hpp>
+#include <ginkgo/core/matrix/coo.hpp>
+#include <ginkgo/core/matrix/csr.hpp>
+#include <ginkgo/core/matrix/diagonal.hpp>
+#include <ginkgo/core/matrix/ell.hpp>
+#include <ginkgo/core/matrix/hybrid.hpp>
+#include <ginkgo/core/matrix/sellp.hpp>
+#include <ginkgo/core/matrix/sparsity_csr.hpp>*/
+
+
+#include "core/components/fill_array.hpp"
+#include "core/matrix/csr_kernels.hpp"
+#include "core/test/utils.hpp"
+#include "test/utils/executor.hpp"
+
+
+namespace {
+
+
+class Csr : public ::testing::Test {
+protected:
+    using itype = int;
+#if GINKGO_COMMON_SINGLE_MODE
+    using vtype = float;
+#else
+    using vtype = double;
+#endif
+    using Mtx = gko::matrix::Csr<vtype, itype>;
+    using Vec = gko::matrix::Dense<vtype>;
+
+    Csr() : rand_engine(15) {}
+
+    void SetUp()
+    {
+        ref = gko::ReferenceExecutor::create();
+        init_executor(ref, exec);
+    }
+
+    void TearDown()
+    {
+        if (exec != nullptr) {
+            ASSERT_NO_THROW(exec->synchronize());
+        }
+    }
+
+    template <typename MtxType>
+    std::unique_ptr<MtxType> gen_mtx(int num_rows, int num_cols)
+    {
+        return gko::test::generate_random_matrix<MtxType>(
+            num_rows, num_cols,
+            std::uniform_int_distribution<>(num_cols, num_cols),
+            std::normal_distribution<>(0.0, 1.0), rand_engine, ref);
+    }
+
+    void set_up_apply_data()
+    {
+        x = gen_mtx<Mtx>(40, 25);
+        alpha = gko::initialize<Vec>({2.0}, ref);
+        dx = Mtx::create(exec);
+        dx->copy_from(x.get());
+        dalpha = Vec::create(exec);
+        dalpha->copy_from(alpha.get());
+    }
+
+    std::shared_ptr<gko::ReferenceExecutor> ref;
+    std::shared_ptr<gko::EXEC_TYPE> exec;
+
+    std::ranlux48 rand_engine;
+
+    std::unique_ptr<Mtx> x;
+    std::unique_ptr<Vec> alpha;
+    std::unique_ptr<Mtx> dx;
+    std::unique_ptr<Vec> dalpha;
+};
+
+
+TEST_F(Csr, ScaleIsEquivalentToRef)
+{
+    set_up_apply_data();
+
+    x->scale(alpha.get());
+    dx->scale(dalpha.get());
+
+    auto result = Mtx::create(ref);
+    result->copy_from(dx.get());
+    GKO_ASSERT_MTX_NEAR(result, x, r<vtype>::value);
+}
+
+
+TEST_F(Csr, InvScaleIsEquivalentToRef)
+{
+    set_up_apply_data();
+
+    x->inv_scale(alpha.get());
+    dx->inv_scale(dalpha.get());
+
+    auto result = Mtx::create(ref);
+    result->copy_from(dx.get());
+    GKO_ASSERT_MTX_NEAR(result, x, r<vtype>::value);
+}
+
+
+}  //  namespace
```

Comment:
nit: again perhaps a new line?
```suggestion
}  //  namespace

```